design/StableRegions-20230209.md

```diff
@@ -0,0 +1,172 @@
+the maximum number of regions and region blocks are each relatively
+small (32k each).  Consequently, they can each be identified with a
+2-byte identifier, and we can pre-allocate tables to store certain
+relations about them, which is critical.
+
+### Q: Are 32767 regions enough?
+
+A: Permitting more than 32k regions may seem theoretically
+interesting, but is not practical given other parameters today that
+control the minimal footprint of a region (8MB) and dictate the
+maximum size of stable memory for a canister today (32GB).  With 32k
+regions at 8MB each, well over the maximum stable memory size is used
+(256GB compared to 32GB, the limit today)
+
+
+## Design details
+
+### API
+
+Internal region allocator operations:
+
+
+ - `initialize` -- called by the RTS, not by the Motoko developer.
+ - `rebuild` -- also called by the RTS, during an post-upgrade.
+
+
+User-facing region allocator operations:
+
+ - `region_new` -- create a dynamic region.
+ - `region_grow` -- grow region by a specified number of pages.
+ - `region_release`* -- release region and reuse region's page blocks.
+ - `region_load` -- read some data from the region.
+ - `region_store` -- store some data into the region.
+
+
+ - * The `region_release` operation is not part of the MVP, but supporting it
+   is important because it means we can transition quickly to an integration
+   with the ambient Motoko GC if we can support it.
+
+## Internal footprint
+
+The state of the allocator is stored in a combination of stable memory and heap memory.
+
+The stable memory state is sufficient to reconstitute the heap objects
+(`rebuild` operation, described in a subsection below).  That means
+that even if the heap is lost (like during an upgrade that fails) the
+stable memory state can fully describe the state parts that should be
+rebuilt into the heap when the upgrade succeeds.
+
+### stable memory fields
+
+ - total allocated buckets, `Nat16`, max value is `32768`.
+ - total allocated regions, `Nat16`, max value is `32767` (one region is reserved for "no region").
+ - bucket-region table (fixed size, about 131kb).
+ - region-object table (fixed size, about 393kb).
+
+### bucket-region table
+
+ - purpose: relate a bucket ID ("page block ID") to its region (if any) and its position in that region.
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BucketRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding bucket ID.
+
+### region-object table
+
+ - 32768 entries (statically sized).
+ - 12 bytes per entry.
+ - entry type = `RegionObjectEntry { size_in_pages: Nat64; heap_object: Nat32 }`
+ - the location of each entry gives its corresponding region ID.
+ - the heap object layout is rebuilt on upgrade.
+ - the heap object is described in its own subsection.
+
+### region heap object
+
+ - `RegionObject { id : Nat16; capacity : Nat16; vec_ptr: Nat32 }`
```

Comment:
> I wonder if you want this here too for easy growth, otherwise you need to consult stable memory to see which entries in the vector are valid (given the spare capacity).

Ah, yes.  We want all accesses to be guarded, so I agree it makes sense to duplicate the `size_in_pages` in the heap object.

> Of these fields, I'd expect all but id to be mutable and change over time, right, as the user grows the region by pages and underlying blocks?

Yes, exactly.  Though, `vec_ptr` and `capacity` will only change (to grow) when the `size_in_pages` doubles.  

(Otherwise, there's a "quadratic foot gun" lurking in the region growth routine for silly user code that grows over and over, by one page at a time.)